libraries/AP_LocationDB/AP_LocationDB.h

```diff
@@ -0,0 +1,141 @@
+#pragma once
+
+#include "AP_LocationDB_config.h"
+
+#if AP_LOCATIONDB_ENABLED
+
+#include <AP_Common/AP_Common.h>
+#include <AP_Common/Location.h>
+#include <AP_Math/AP_Math.h>
+#include <AP_HAL/AP_HAL_Boards.h>
+#include <AP_HAL/Semaphores.h>
+#include <AP_Vehicle/AP_Vehicle_Type.h>
+#include <AC_Avoidance/AP_OABendyRuler.h>
+
+#define AP_LOCATIONDB_ITEM_SIZE 56
+
+class AP_LocationDB_Item {
+public:
+    friend class AP_Filesystem_LocationDB;
+    friend class AP_LocationDB;
+
+    AP_LocationDB_Item() {};
+    AP_LocationDB_Item(uint32_t key, uint32_t timestamp_ms, Vector3f pos_cm_NEU, Vector3f vel_cm_NEU, Vector3f acc_cm_NEU, float heading_cdeg, float radius_cm, uint8_t populated_fields);
+
+    // to initialise an empty database item with given data
+    void init(uint32_t key, uint32_t timestamp_ms, Vector3f pos_cm_NEU, Vector3f vel_cm_NEU, Vector3f acc_cm_NEU, float heading_cdeg, float radius_cm, uint8_t populated_fields);
+    
+    uint32_t get_timestamp_ms() { return _timestamp_ms; }
+    uint32_t get_key() { return _key; }
+    bool get_pos_cm_NEU(Vector3f &ret) const;
+    bool get_vel_cm_NEU(Vector3f &ret) const;
+    bool get_acc_cm_NEU(Vector3f &ret) const;
+    bool get_heading_cdeg(float &heading) const;
+    bool get_radius_cm(float &ret) const;
+
+    enum class DataField : uint8_t {
+        POS = (1U << 0),
+        VEL = (1U << 1),
+        ACC = (1U << 2),
+        HEADING = (1U << 3),
+        RADIUS = (1U << 4),
+    };
+
+    enum class Flag : uint8_t {
+        AVOID = (1U << 0),
+    };
+
+private:
+    uint32_t _key;
+    uint32_t _timestamp_ms;
+    Vector3f _pos;
+    Vector3f _vel;
+    Vector3f _acc;
+    float _heading;
+    float _radius;
+    uint8_t _flags;
+
+    uint8_t _populated_fields = 0;
+
+    bool is_field_populated(DataField option) const {
+        return (_populated_fields & (uint8_t)option) != 0;
+    }
+};
+
+class AP_LocationDB {
+public:
+    friend class AP_Filesystem_LocationDB;
+    friend class AP_OABendyRuler;
```

Comment:
I'm wondering why Bendy Ruler should be a friend class? Should we not try to avoid entangling them?

Reply:
We do not give sequential access to the database elements to the outside world. To access any element, we need to have the key to the element. This works good in most of the cases. AP_OABendyRuler needs to access all the elements sequentially and hence I have made AP_OABendyRuler a friend class so that it can access the private method `get_item_at_index`

Reply:
It would be better IMO to provide an iterator that can allow Bendy Ruler (or any consumer) to access the elements sequentially.